File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/http/CustomHttpServiceInterceptor.kt
package com.mapbox.maps.mapbox_maps.http

import android.os.Handler
import android.os.Looper
import com.mapbox.common.HttpRequest
import com.mapbox.common.HttpRequestOrResponse
import com.mapbox.common.HttpResponse
import com.mapbox.common.HttpServiceFactory
import com.mapbox.common.HttpServiceInterceptorInterface
import com.mapbox.common.HttpServiceInterceptorRequestContinuation
import com.mapbox.common.HttpServiceInterceptorResponseContinuation
import io.flutter.plugin.common.MethodChannel

class CustomHttpServiceInterceptor : HttpServiceInterceptorInterface {
  // Thread synchronization lock for protecting mutable state
  private val lock = Any()

  private var customHeaders: MutableMap<String, String> = mutableMapOf()
  private var flutterChannel: MethodChannel? = null
  private var interceptRequests: Boolean = false
  private var interceptResponses: Boolean = false
  private var includeResponseBody: Boolean = false
  private val mainHandler = Handler(Looper.getMainLooper())

  override fun onRequest(request: HttpRequest, continuation: HttpServiceInterceptorRequestContinuation) {
    val currentHeaders = HashMap(request.headers)

    // Thread-safe access to mutable state
    val (headers, shouldIntercept, channel) = synchronized(lock) {
      Triple(HashMap(customHeaders), interceptRequests, flutterChannel)
    }

    // First apply static custom headers
    currentHeaders.putAll(headers)

    // If Flutter callback is enabled, invoke it
    if (shouldIntercept && channel != null) {
      invokeFlutterRequestCallback(request, currentHeaders, channel, continuation)
    } else {
      // No callback, just continue with static headers
      val modifiedRequest = request.toBuilder()
        .headers(currentHeaders)
        .build()
      val requestOrResponse = HttpRequestOrResponse(modifiedRequest)
      continuation.run(requestOrResponse)
    }
  }

  private fun invokeFlutterRequestCallback(
    request: HttpRequest,
    currentHeaders: HashMap<String, String>,
    channel: MethodChannel,
    continuation: HttpServiceInterceptorRequestContinuation
  ) {
    val requestMap = mapOf(
      "url" to request.url,
      "method" to request.method.name,
      "headers" to currentHeaders,
      "body" to request.body
    )

    // Use async continuation pattern - call continuation directly from the Flutter callback
    // This avoids blocking the calling thread and prevents deadlocks when called from main thread
    mainHandler.post {
      channel.invokeMethod("http#onRequest", requestMap, object : MethodChannel.Result {
        override fun success(result: Any?) {
          val builder = request.toBuilder()

          @Suppress("UNCHECKED_CAST")
          val modified = result as? Map<String, Any?>

          if (modified != null) {
            // Apply modified headers
            @Suppress("UNCHECKED_CAST")
            val newHeaders = modified["headers"] as? Map<String, String>
            if (newHeaders != null) {
              builder.headers(HashMap(newHeaders))
            } else {
              builder.headers(currentHeaders)
            }

            // Apply modified URL
            val newUrl = modified["url"] as? String
            if (newUrl != null) {
              builder.url(newUrl)
            }

            // Apply modified body
            @Suppress("UNCHECKED_CAST")
            val newBody = modified["body"] as? ByteArray
            if (newBody != null) {
              builder.body(newBody)
            }
          } else {
            builder.headers(currentHeaders)
          }

          val modifiedRequest = builder.build()
          val requestOrResponse = HttpRequestOrResponse(modifiedRequest)
          // Call continuation from inside the callback - this is safe because
          // the native backend request() call is async (queues the request and returns immediately)
          continuation.run(requestOrResponse)
        }

        override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
          // On error, continue with original request + custom headers
          val modifiedRequest = request.toBuilder()
            .headers(currentHeaders)
            .build()
          val requestOrResponse = HttpRequestOrResponse(modifiedRequest)
          continuation.run(requestOrResponse)
        }

        override fun notImplemented() {
          // Not implemented, continue with original request + custom headers
          val modifiedRequest = request.toBuilder()
            .headers(currentHeaders)
            .build()
          val requestOrResponse = HttpRequestOrResponse(modifiedRequest)
          continuation.run(requestOrResponse)
        }
      })
    }
  }

  override fun onResponse(response: HttpResponse, continuation: HttpServiceInterceptorResponseContinuation) {
    // Thread-safe access to mutable state
    val (shouldIntercept, channel, shouldIncludeBody) = synchronized(lock) {
      Triple(interceptResponses, flutterChannel, includeResponseBody)
    }

    // If Flutter callback is enabled, invoke it (non-blocking)
    if (shouldIntercept && channel != null) {
      // response.result is Expected<HttpRequestError, HttpResponseData>
      // We need to check if it's a success and extract the value
      val responseData = response.result.value
      val responseMap = if (responseData != null) {
        mapOf(
          "url" to response.request.url,
          "statusCode" to responseData.code,
          "headers" to responseData.headers,
          // Only include response body if explicitly opted in (to avoid performance issues)
          "data" to if (shouldIncludeBody) responseData.data else null,
          "requestHeaders" to response.request.headers
        )
      } else {
        // Error response - still send basic info
        mapOf(
          "url" to response.request.url,
          "statusCode" to -1,
          "headers" to emptyMap<String, String>(),
          "data" to null,
          "requestHeaders" to response.request.headers
        )
      }

      mainHandler.post {
        channel.invokeMethod("http#onResponse", responseMap, object : MethodChannel.Result {
          override fun success(result: Any?) {}
          override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {}
          override fun notImplemented() {}
        })
      }
    }

    continuation.run(response)
  }

  fun setCustomHeaders(headers: Map<String, String>) {
    synchronized(lock) {
      customHeaders.clear()
      customHeaders.putAll(headers)
    }
    HttpServiceFactory.setHttpServiceInterceptor(this)
  }

  fun setFlutterChannel(channel: MethodChannel?) {
    synchronized(lock) {
      flutterChannel = channel
    }
  }

  fun setInterceptorEnabled(
    enabled: Boolean,
    interceptRequests: Boolean,
    interceptResponses: Boolean,
    includeResponseBody: Boolean = false
  ) {
    synchronized(lock) {
      this.interceptRequests = enabled && interceptRequests
      this.interceptResponses = enabled && interceptResponses
      this.includeResponseBody = includeResponseBody
    }
    HttpServiceFactory.setHttpServiceInterceptor(this)
  }

  companion object {
    @Volatile
    private var instance: CustomHttpServiceInterceptor? = null

    fun getInstance(): CustomHttpServiceInterceptor {
      return instance ?: synchronized(this) {
        instance ?: CustomHttpServiceInterceptor().also { instance = it }
      }
    }
  }
}

File: ios/mapbox_maps_flutter/Sources/mapbox_maps_flutter/Classes/CustomHttpServiceInterceptor.swift
import Flutter
import MapboxMaps

final class CustomHttpServiceInterceptor: HttpServiceInterceptorInterface {
    static let shared = CustomHttpServiceInterceptor()

    // Thread synchronization lock for protecting mutable state
    private let lock = NSLock()

    private var _customHeaders: [String: String] = [:]
    private var _flutterChannel: FlutterMethodChannel?
    private var _interceptRequests: Bool = false
    private var _interceptResponses: Bool = false
    private var _includeResponseBody: Bool = false

    var customHeaders: [String: String] {
        get { lock.withLock { _customHeaders } }
        set { lock.withLock { _customHeaders = newValue } }
    }

    var flutterChannel: FlutterMethodChannel? {
        get { lock.withLock { _flutterChannel } }
        set { lock.withLock { _flutterChannel = newValue } }
    }

    var interceptRequests: Bool {
        get { lock.withLock { _interceptRequests } }
        set { lock.withLock { _interceptRequests = newValue } }
    }

    var interceptResponses: Bool {
        get { lock.withLock { _interceptResponses } }
        set { lock.withLock { _interceptResponses = newValue } }
    }

    var includeResponseBody: Bool {
        get { lock.withLock { _includeResponseBody } }
        set { lock.withLock { _includeResponseBody = newValue } }
    }

    private init() {}

    func onRequest(
        for request: HttpRequest, continuation: @escaping HttpServiceInterceptorRequestContinuation
    ) {
        var modifiedHeaders = request.headers

        // First apply static custom headers
        for (key, value) in customHeaders {
            modifiedHeaders[key] = value
        }

        // If Flutter callback is enabled, invoke it
        if interceptRequests, let channel = flutterChannel {
            invokeFlutterRequestCallback(
                request: request,
                currentHeaders: modifiedHeaders,
                channel: channel,
                continuation: continuation
            )
        } else {
            // No callback, just continue with static headers
            let modifiedRequest = HttpRequest(
                method: request.method,
                url: request.url,
                headers: modifiedHeaders,
                timeout: request.timeout,
                networkRestriction: request.networkRestriction,
                sdkInformation: request.sdkInformation,
                body: request.body,
                flags: request.flags
            )
            continuation(HttpRequestOrResponse.fromHttpRequest(modifiedRequest))
        }
    }

    private func invokeFlutterRequestCallback(
        request: HttpRequest,
        currentHeaders: [String: String],
        channel: FlutterMethodChannel,
        continuation: @escaping HttpServiceInterceptorRequestContinuation
    ) {
        let requestMap: [String: Any?] = [
            "url": request.url,
            "method": request.method.rawValue,
            "headers": currentHeaders,
            "body": request.body,
        ]

        // Use async continuation pattern - call continuation directly from the Flutter callback
        // This avoids blocking the calling thread and prevents deadlocks when called from main thread
        DispatchQueue.main.async {
            channel.invokeMethod("http#onRequest", arguments: requestMap) { result in
                var finalHeaders = currentHeaders
                var finalUrl = request.url
                var finalBody = request.body

                if let modified = result as? [String: Any] {
                    // Apply modified headers
                    if let newHeaders = modified["headers"] as? [String: String] {
                        finalHeaders = newHeaders
                    }

                    // Apply modified URL
                    if let newUrl = modified["url"] as? String {
                        finalUrl = newUrl
                    }

                    // Apply modified body
                    if let newBody = modified["body"] as? FlutterStandardTypedData {
                        finalBody = newBody.data
                    } else if let newBody = modified["body"] as? Data {
                        finalBody = newBody
                    }
                }

                let modifiedRequest = HttpRequest(
                    method: request.method,
                    url: finalUrl,
                    headers: finalHeaders,
                    timeout: request.timeout,
                    networkRestriction: request.networkRestriction,
                    sdkInformation: request.sdkInformation,
                    body: finalBody,
                    flags: request.flags
                )
                // Call continuation from inside the callback - this is safe because
                // the native backend request() call is async (queues the request and returns immediately)
                continuation(HttpRequestOrResponse.fromHttpRequest(modifiedRequest))
            }
        }
    }

    func onResponse(
        for response: HttpResponse,
        continuation: @escaping HttpServiceInterceptorResponseContinuation
    ) {
        // If Flutter callback is enabled, invoke it (non-blocking)
        if interceptResponses, let channel = flutterChannel {
            // Capture includeResponseBody value under lock to ensure thread safety
            let shouldIncludeBody = includeResponseBody

            // response.result is Result<HttpResponseData, HttpRequestError>
            // We need to extract the success value
            var responseMap: [String: Any?]
            switch response.result {
            case .success(let responseData):
                responseMap = [
                    "url": response.request.url,
                    "statusCode": Int(responseData.code),
                    "headers": responseData.headers,
                    // Only include response body if explicitly opted in (to avoid performance issues)
                    "data": shouldIncludeBody ? responseData.data : nil,
                    "requestHeaders": response.request.headers,
                ]
            case .failure(_):
                responseMap = [
                    "url": response.request.url,
                    "statusCode": -1,
                    "headers": [:] as [String: String],
                    "data": nil,
                    "requestHeaders": response.request.headers,
                ]
            }

            DispatchQueue.main.async {
                channel.invokeMethod("http#onResponse", arguments: responseMap, result: nil)
            }
        }

        continuation(response)
    }

    func setFlutterChannel(_ channel: FlutterMethodChannel?) {
        flutterChannel = channel
    }

    func setInterceptorEnabled(
        enabled: Bool,
        interceptRequests: Bool,
        interceptResponses: Bool,
        includeResponseBody: Bool = false
    ) {
        self.interceptRequests = enabled && interceptRequests
        self.interceptResponses = enabled && interceptResponses
        self.includeResponseBody = includeResponseBody
        HttpServiceFactory.setHttpServiceInterceptorForInterceptor(self)
    }
}
